### ISSUE TITLE
pgcat as an alternative to pgbouncer and more

### ISSUE THREAD

**Issue by rcknr** [vitabaks/postgresql_cluster#646]
I have recently deployed a cluster using this playbook and am very pleased with the outcome. As I learn capabilities of each component I also keep discovering new tools that can enhance database performance in my production scenario.
Recently I came across [pgcat](https://postgresml.org/docs/product/pgcat/features) which is designed to be a replacement for pgbouncer but also could be a way to natively balance reads/writes as well as sharded datasets.
For now I'm testing it in parallel with haproxy/pgbouncer in my cluster populating members using confd/etcd.
Do you think it makes sense to include it into this playbook as an alternative to pgbouncer and haproxy? Does anyone have any experience working with pgcat and can share it here?

**Comment by vitabaks** [comment]
Yes, I have an idea to add **pgcat** as an alternative to pgbouncer so that the user can choose which connection pooler to use. 

I need to take the time to do this, or you can suggest a PR.

**Comment by rcknr** [comment]
Currently, the playbook installs a copy of pgbouncer to each cluster node. With pgcat I currently test it with a single instance replacing haproxy as well. If you install pgcat instead of pgbouncer which topology would you use?

**Comment by vitabaks** [comment]
In the same configuration

by default, it is placed locally on the database server before Postgres.

**Comment by TheOriginalGraLargeShrimpakaReaper** [comment]
I opened an Discussion on PgCat.
My Idea was, that PgCat replaced HAproxy and PgBouncer because it is a Proxy and Loud Balancer with SQL Inspect for Splitting Read/Write.
The Problem is, it seems that PgCat is not able to Check the Patroni State.

@vitabaks what did you find out about this?

**Comment by rcknr** [comment]
@TheOriginalGraLargeShrimpakaReaper why do you think it is a problem? pgcat claims to have its own healthcheck performed by connecting to PostgreSQL directly. Cluster configuration can be populated from etcd with confd the same way HAProxy is configured by this playbook. pgcat even reloads its configuration on its own.

**Comment by vitabaks** [comment]
> My Idea was, that PgCat replaced HAproxy and PgBouncer because it is a Proxy and Loud Balancer with SQL Inspect for Splitting Read/Write.
The Problem is, it seems that PgCat is not able to Check the Patroni State.

In this case, you need to understand that it will have fewer features than a load balancer integrated with the Patroni REST API. For example, we can now define sync and async replicas separately, automatically exclude replicas with a lag of more than `patroni_maximum_lag_on_replica` and use user tags (`balancer_tags`), for example, to be able to split read traffic across different datacenters. In the case of pgcat, as I understand it, this will be the usual load balancing for reading, but perhaps this will be enough for many.

By default, HAProxy and PgBouncer will be specified, but those who know what they are doing (understand the limitations) will be able to select pgcat to use its functions.

If we consider it not only as a connection pooler but also as a load balancer, then its servers should be defined in the "balancers" group to be able to control which servers to deploy it on (dedicated or database servers).

**Comment by vitabaks** [comment]
@rcknr Ok, thank you for your feedback. We will be watching the next releases.

**Comment by vitabaks** [comment]
If you're interested in this feature, please consider becoming a [sponsor](https://autobase.tech/docs/sponsor).

The development of this feature requires sponsorship to fund developer efforts. If you're already a sponsor and are interested in this feature, please leave a comment here so we can prioritize this issue accordingly.

We also welcome contributions from those willing to dedicate their time to implement this feature on a voluntary basis. Please see the [Contributing Guide](https://github.com/vitabaks/postgresql_cluster/blob/master/CONTRIBUTING.md) for more information.

**Comment by pat-s** [comment]
I've read the discussion here with great interest. `pgcat` has meanwhile seen a few more releases and is at v0.2.5. OTOH, there have only been ~ 10 commits in the last 4 months.

**Comment by vitabaks** [comment]
Another alternative with the possibility of sharding

https://github.com/pgdogdev/pgdog

**Comment by vitabaks** [comment]
Spqr https://github.com/pg-sharding/spqr

**Comment by vitabaks** [comment]
Multigres: https://github.com/multigres/multigres

Info: https://supabase.com/blog/multigres-vitess-for-postgres

I’m planning to use **pgdog** as the default LB/pooler/shard-proxy in Autobase v3.0.
However, if **multigres** proves to be solid, we can also include it as an alternative — to give users more flexibility and choice.

**Comment by skauk** [comment]
In the meantime, pgcat has seen some development and matured over v1. I'll have a look when I have a chance.

**Comment by vitabaks** [comment]
pgdog is a continuation of pgcat development, offering more features and a stronger focus on sharding. I don’t believe pgcat is worth considering at this point.

**Comment by shamrin** [comment]
@vitabaks FYI pgdog is AGPL-licensed, since January 2025: https://github.com/pgdogdev/pgdog/blob/main/LICENSE

**Comment by shamrin** [comment]
There is also odyssey: https://github.com/yandex/odyssey

**Comment by vitabaks** [comment]
@shamrin I don't think this is bothering us, because we are not a cloud service, we just install and configure services to your infrastructure.

> Autobase installs AGPL-licensed components (e.g., pgdog) as separate unmodified packages inside the customer’s infrastructure. Autobase does not incorporate or distribute AGPL code, nor provide network service access to it, therefore AGPL copyleft obligations do not apply.

**Comment by vitabaks** [comment]
> There is also odyssey

Yes, we’re aware of it, and we even use it in some places. But pgdog is more than just a connection pooler, so we see it as a potential replacement for both HAProxy and PgBouncer.
It will also provide sharding capabilities, enabling Autobase to manage much larger database clusters.

**Comment by shamrin** [comment]
> Autobase installs AGPL-licensed components (e.g., pgdog) as separate unmodified packages

@vitabaks yes I understand. I have raised AGPL issue because some companies have strict policies against it. So far all software autobase suggests is permissively licensed. Having something that is not may be breaking unspoken expectations. If pgdog is optional and clearly marked as “AGPL, beware”, then it is okay.

**Comment by pat-s** [comment]
I've lately did some research about pgdog, including watching [this video](https://youtu.be/3ARKt62iVlA?is=xJsAsbcw9TLVs6FH) from the creator. 

[This comparison table](https://docs.pgdog.dev/architecture/comparison/#basics) is also interesting.

Project status sounds as if it should be safe to integrate:

> PgDog is used in production and at scale. Most features are stable, while some are experimental. Check [documentation](https://docs.pgdog.dev/features/) for more details. New sharding features are added almost weekly.

**Comment by vitabaks** [comment]
We are already actively [testing](https://github.com/pgdogdev/pgdog/issues?q=is%3Aissue%20author%3Avitabaks) pgdog and helping to identify bugs. So I think we will soon be developing a 3.0 release in which we will replace the current balancing components with pgdog proxy.

**Comment by iRazvan2745** [comment]
Hey, any news on this?

**Comment by vitabaks** [comment]
> Hey, any news on this?

We are currently focused on finalizing the platform UI. I think until the end of this year.

We are also waiting for the first production-ready release of pgdog.
